File: iosApp/iosApp/Presentation/Screens/UserList/UserListViewModel.swift
import Foundation
import Combine
import Shared

/// ViewModel for the user list screen
/// Replicates functionality from Kotlin UserListViewModel
@MainActor
class UserListViewModel: ObservableObject {
    @Published private(set) var state: UserListState = .initial
    @Published var searchQuery: String = "" {
        didSet {
            onSearchQueryChange(query: searchQuery)
        }
    }
    
    private let getUsersUseCase: GetUsersUseCase
    private var cancellables = Set<AnyCancellable>()
    
    init(getUsersUseCase: GetUsersUseCase? = nil) {
        // Use injected use case or get from Koin
        self.getUsersUseCase = getUsersUseCase ?? DependencyContainer.shared.getUsersUseCase()
        loadUsers()
    }
    
    /// Load users from the API
    func loadUsers() {
        state = .loading
        
        Task {
            await fetchUsers()
        }
    }
    
    /// Refresh users (for pull-to-refresh)
    func refresh() async {
        // Set refreshing state while keeping current data visible
        if case let .success(users, filteredUsers, _) = state {
            state = .success(users: users, filteredUsers: filteredUsers, isRefreshing: true)
        }
        
        await fetchUsers()
    }
    
    /// Fetch users from the use case
    private func fetchUsers() async {
        do {
            let result = try await getUsersUseCase.invoke(count: 20)
            
            handleNetworkResult(result)
        } catch {
            state = .error(message: "An unexpected error occurred: \(error.localizedDescription)")
        }
    }
    
    /// Handle network result from use case
    private func handleNetworkResult(_ result: NetworkResult<NSArray>) {
        if let successResult = result as? NetworkResultSuccess<NSArray>,
           let data = successResult.data {
            let users = data.compactMap { $0 as? User }
            let currentQuery = searchQuery
            let filteredUsers = currentQuery.isEmpty ? users : users.filter { user in
                user.fullName.localizedCaseInsensitiveContains(currentQuery)
            }
            state = .success(users: users, filteredUsers: filteredUsers, isRefreshing: false)
        } else if let httpError = result as? NetworkResultHttpError {
            state = .error(message: "Server error: \(httpError.code)")
        } else if result is NetworkResultNetworkError {
            state = .error(message: "Network error. Please check your connection.")
        } else if result is NetworkResultSerializationError {
            state = .error(message: "Data parsing error")
        } else if result is NetworkResultUnknownError {
            state = .error(message: "An unexpected error occurred")
        } else {
            state = .error(message: "An unexpected error occurred")
        }
    }
    
    /// Update search query and filter users
    private func onSearchQueryChange(query: String) {
        if case let .success(users, _, _) = state {
            let filtered = query.isEmpty ? users : users.filter { user in
                user.fullName.localizedCaseInsensitiveContains(query)
            }
            state = .success(users: users, filteredUsers: filtered, isRefreshing: false)
        }
    }
}

File: composeApp/src/commonMain/kotlin/com/example/coveragex/presentation/screens/userlist/UserListViewModel.kt
package com.example.coveragex.presentation.screens.userlist

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.coveragex.domain.usecase.GetUsersUseCase
import com.example.coveragex.util.NetworkResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for the user list screen.
 */
class UserListViewModel(
    private val getUsersUseCase: GetUsersUseCase
) : ViewModel() {

    private val _state = MutableStateFlow<UserListState>(UserListState.Initial)
    val state: StateFlow<UserListState> = _state.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    init {
        loadUsers()
    }

    /**
     * Load users from the API
     */
    fun loadUsers() {
        viewModelScope.launch {
            _state.value = UserListState.Loading
            fetchUsers()
        }
    }

    /**
     * Refresh users (for pull-to-refresh)
     */
    fun refresh() {
        viewModelScope.launch {
            // Set refreshing state while keeping current data visible
            val currentState = _state.value
            if (currentState is UserListState.Success) {
                _state.value = currentState.copy(isRefreshing = true)
            }
            
            fetchUsers()
        }
    }

    /**
     * Fetch users from the use case
     */
    private suspend fun fetchUsers() {
        when (val result = getUsersUseCase(count = 20)) {
            is NetworkResult.Success -> {
                val currentQuery = _searchQuery.value
                val filteredUsers = if (currentQuery.isBlank()) {
                    result.data
                } else {
                    result.data.filter { user ->
                        user.fullName.contains(currentQuery, ignoreCase = true)
                    }
                }
                
                _state.value = UserListState.Success(
                    users = result.data,
                    filteredUsers = filteredUsers,
                    isRefreshing = false
                )
            }
            is NetworkResult.HttpError -> {
                _state.value = UserListState.Error(
                    "Server error: ${result.code}"
                )
            }
            is NetworkResult.NetworkError -> {
                _state.value = UserListState.Error(
                    "Network error. Please check your connection."
                )
            }
            is NetworkResult.SerializationError -> {
                _state.value = UserListState.Error(
                    "Data parsing error"
                )
            }
            is NetworkResult.UnknownError -> {
                _state.value = UserListState.Error(
                    "An unexpected error occurred"
                )
            }
        }
    }

    /**
     * Update search query and filter users
     */
    fun onSearchQueryChange(query: String) {
        _searchQuery.value = query

        val currentState = _state.value
        if (currentState is UserListState.Success) {
            val filtered = if (query.isBlank()) {
                currentState.users
            } else {
                currentState.users.filter { user ->
                    user.fullName.contains(query, ignoreCase = true)
                }
            }

            _state.value = currentState.copy(filteredUsers = filtered)
        }
    }


}
